exercises/palindrome-products/palindrome_products_test.py

```diff
@@ -16,31 +16,66 @@
 from palindrome_products import smallest_palindrome, largest_palindrome
 
 
+# Tests adapted from `problem-specifications//canonical-data.json` @ v1.0.0
+
 class PalindromesTests(unittest.TestCase):
+    def test_smallest_palindrome_from_single_digit_factors(self):
+        value, factors = smallest_palindrome(min_factor=1, max_factor=9)
+        self.assertEqual(value, 1)
+        self.assertEqual(factors, {frozenset((1, 1))})
+
     def test_largest_palindrome_from_single_digit_factors(self):
-        value, factors = largest_palindrome(max_factor=9)
+        value, factors = largest_palindrome(min_factor=1, max_factor=9)
         self.assertEqual(value, 9)
-        self.assertIn(set(factors), [{1, 9}, {3, 3}])
+        self.assertEqual(factors, {frozenset((1, 9)), frozenset((3, 3))})
```

Comment:
Not sure I agree with the usage of nested sets here... the canonical data expects each element of `factors` to be a pair (tuple, list, etc), and returning as a set makes duplicate factor pairs into singletons. I would suggest the following form instead:
```Python
expected = {(1, 9), (3, 3)}
```

Reply:
Sets were used to be able to compare individual tuples without having to worry about the orders of their elements, so, for example, `(1, 9)` and `(9, 1)` would be equivalent.

An alternative may be to sort the elements of each tuple in the set, what are your thoughts on that?

Reply:
There's precedence for sorting tuple elements, actually: see the [`normalize_dominoes` function in the `dominoes` exercise](https://github.com/exercism/python/blob/e78fd8039e12ff24e4efff689fed31e7e7cbf4da/exercises/dominoes/dominoes_test.py#L72-L73). It may not be exactly what you need but would get you pretty close. That function also would catch duplicate pairs, since it works with lists instead of sets, but modifying it to use sets wouldn't be hard.